File: 0069_Sqrt_x/main.cpp
// T = O(logx), S = O(1)
class Solution {
  public:
    int mySqrt(int x) {
        long l = 0, r = x;
        while (l < r) {
            int mid = (l + r + 1) >> 1;
            if (mid <= x / mid) {
                l = mid;
            } else {
                r = mid - 1;
            }
        }
        return l;
    }
};

// T = O(logx), S = O(1)
class Solution {
  public:
    int mySqrt(int x) {
        long n = x;
        while (n * n > x) {
            n = (n + x / n) >> 1;
        }
        return n;
    }
};

File: 0069_Sqrt_x/golang/main.go
// T = O(logx), S = O(1)
func mySqrt(x int) int {
	l, r := 0, x
	for l < r {
		mid := l + (r-l+1)/2
		if mid <= x/mid {
			l = mid
		} else {
			r = mid - 1
		}
	}
	return l
}